Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift

```diff
@@ -441,6 +474,42 @@ class HTTPClientUpgradeTestCase: XCTestCase {
             .assertDoesNotContain(handlerType: NIOHTTPClientUpgradeHandler.self))
     }
     
+    func testUpgradeCompleteFlush() throws {
+        var upgradeHandlerCallbackFired = false
+        let handler = ChannelReadWriteHandler()
+        let upgrader = AddHandlerClientUpgrader(forProtocol: "myproto", addingHandler: handler)
+        let clientChannel = try setUpClientChannel(clientHTTPHandler: ExplodingHTTPHandler(),
+                                                   clientUpgraders: [upgrader]) { (context) in
+                                                    
+                                                    upgradeHandlerCallbackFired = true
+        }
+        defer {
+            XCTAssertNoThrow(try clientChannel.finish())
+        }
+        
+        // Read the server request.
+        if let requestString = try clientChannel.readByteBufferOutputAsString() {
+                        
+            XCTAssertEqual(requestString, "GET / HTTP/1.1\r\nContent-Type: text/plain; charset=utf-8\r\nContent-Length: 0\r\nConnection: upgrade\r\nUpgrade: myproto\r\n\r\n")
+        } else {
+            XCTFail()
+        }
+        
+        // Push the successful server response.
+        let response = "HTTP/1.1 101 Switching Protocols\r\nConnection: upgrade\r\nUpgrade: myproto\r\n\r\nTest"
+                
+        XCTAssertNoThrow(try clientChannel.writeInbound(ByteBuffer.forString(response)))
+        
+        clientChannel.embeddedEventLoop.run()
+                
+        XCTAssert(upgradeHandlerCallbackFired)
+        
+        XCTAssertEqual(handler.messagesReceived, 1)
+        
+        XCTAssertNoThrow(try clientChannel.pipeline
+            .assertDoesNotContain(handlerType: NIOHTTPClientUpgradeHandler.self))
```

Comment:
Can you also do a `readByteBufferOutputAsString` and assert it contains the server response we wrote?

Reply:
I tried adding this at the end, but the result is `nil`. However the `ByteBuffer` I get in my `ChannelReadWriteHandler` does contain the string I put in the response. Is there a step I'm skipping to have said buffer end up at the end. The pipeline looks like this:
```
ChannelPipeline[ObjectIdentifier(0x0000000103503a50)]:
                     [I] ↓↑ [O]
 ChannelReadWriteHandler ↓↑  [handler4]
```

Reply:
You added `clientChannel.readByteBufferOutputAsString`? Whereabouts did you add that line?

Reply:
Right at the end. I can push the non-working version

Reply:
Please do.

Reply:
Pushed! 😄